experimental/lib/Transforms/ResourceSharing/Crush.cpp

```diff
@@ -0,0 +1,682 @@
+        sharingInfo(sharingInfo){};
+  SharingInfo &sharingInfo;
+
+#ifndef DYNAMATIC_GUROBI_NOT_INSTALLED
+  LogicalResult getBufferPlacement(FuncInfo &funcInfo, TimingDatabase &timingDB,
+                                   Logger *logger,
+                                   BufferPlacement &placement) override {
+
+    // Create Gurobi environment
+    GRBEnv env = GRBEnv(true);
+    env.set(GRB_IntParam_OutputFlag, 0);
+    if (timeout > 0)
+      env.set(GRB_DoubleParam_TimeLimit, timeout);
+    env.start();
+
+    if (algorithm == FPGA20 || algorithm == FPGA20_LEGACY) {
+      // Create and solve the MILP
+      return checkLoggerAndSolve<FPGA20BuffersWrapper>(
+          logger, "placement", placement, sharingInfo, env, funcInfo, timingDB,
+          targetCP, algorithm != FPGA20);
+    }
+    if (algorithm == FPL22) {
+      // Create disjoint block unions of all CFDFCs
+      SmallVector<CFDFC *, 8> cfdfcs;
+      std::vector<CFDFCUnion> disjointUnions;
+      llvm::transform(funcInfo.cfdfcs, std::back_inserter(cfdfcs),
+                      [](auto cfAndOpt) { return cfAndOpt.first; });
+      getDisjointBlockUnions(cfdfcs, disjointUnions);
+
+      // Create and solve an MILP for each CFDFC union. Placement decisions get
+      // accumulated over all MILPs. It's not possible to override a previous
+      // placement decision because each CFDFC union is disjoint from the others
+      for (auto [id, cfUnion] : llvm::enumerate(disjointUnions)) {
+        std::string milpName = "cfdfc_placement_" + std::to_string(id);
+        if (failed(checkLoggerAndSolve<FPL22BuffersWraper>(
+                logger, milpName, placement, sharingInfo, env, funcInfo,
+                timingDB, targetCP, cfUnion)))
+          return failure();
+      }
+
+      // Solve last MILP on channels/units that are not part of any CFDFC
+      return checkLoggerAndSolve<fpl22::OutOfCycleBuffers>(
+          logger, "out_of_cycle", placement, env, funcInfo, timingDB, targetCP);
+    }
+
+    llvm_unreachable("unknown algorithm");
+  }
+#endif // DYNAMATIC_GUROBI_NOT_INSTALLED
+};
+
+struct CreditBasedSharingPass
+    : public dynamatic::experimental::sharing::impl::CreditBasedSharingBase<
+          CreditBasedSharingPass> {
+
+  CreditBasedSharingPass(StringRef algorithm, StringRef frequencies,
+                         StringRef timingModels, bool firstCFDFC,
+                         double targetCP, unsigned timeout, bool dumpLogs) {
+    this->algorithm = algorithm.str();
+    this->frequencies = frequencies.str();
+    this->timingModels = timingModels.str();
+    this->firstCFDFC = firstCFDFC;
+    this->targetCP = targetCP;
+    this->timeout = timeout;
+    this->dumpLogs = dumpLogs;
+  }
+
+  void runDynamaticPass() override;
+
+  SmallVector<mlir::Operation *> getSharingTargets(handshake::FuncOp funcOp) {
+    SmallVector<Operation *> sharingTargets;
+
+    for (auto &op : funcOp.getOps()) {
+      if (isa<SHARING_TARGETS>(op)) {
```

Comment:
I am not a big fan of using a macro for this (not really standard, and usually error-prone), why not just write all the type names directly here? Alternatively, create an inlined function `static inline bool isSharingTarget(Operation*)` and call it from here.